crates/cadenza-eval/STATUS.md

```diff
@@ -684,6 +684,98 @@ The following enhancements were identified during code review and are planned fo
+
+**High Priority Tasks** (in order):
+- [x] ~~**Value location tracking and local management**~~  ✅
+  - [x] Design data structure to map ValueId → LocalIndex or StackPosition
+  - [x] Track function parameters as locals 0..N
+  - [x] Allocate locals for SSA values (one local per ValueId that needs storage)
+  - [x] Implement strategy: parameters as locals, results pushed to stack
+  - [x] Generate `local.get` instructions to load operands before operations
+  - [x] Generate `local.set` instructions when values need to be stored
+  - [x] Handle the case where a value is used multiple times (must store in local)
+- [ ] **Function calls**
+  - [ ] Load arguments in order using `local.get` or stack values
+  - [ ] Generate `call` instruction with correct function index
+  - [ ] Handle return values on stack
+  - [ ] Test with recursive functions
+- [ ] **Control flow**
+  - [ ] Implement branch instruction generation (conditional jumps)
+  - [ ] Implement unconditional jump (br) for loops
+  - [ ] Handle block nesting for structured control flow
+  - [ ] Generate proper WASM blocks and loops
+  - [ ] Map IR basic blocks to WASM control structures
+- [ ] **Unary operations**
+  - [ ] Fix negation to properly load operand first
+  - [ ] Fix logical not with proper type conversions
+  - [ ] Implement bitwise not
+
+**Medium Priority** (after basic codegen works):
+- [ ] **String constants**
+  - [ ] Use WASM GC arrays for string representation
+  - [ ] Import/export string handling functions
+  - [ ] Consider WTF-8 encoding for compatibility
+- [ ] **Quantity/measurement values**
+  - [ ] Represent as structs with value and unit fields
+  - [ ] Or compile-time eliminate units and use raw numbers
+- [ ] **Records and tuples**
+  - [ ] Use WASM GC struct types
+  - [ ] Generate field access instructions (struct.get)
+  - [ ] Generate struct construction (struct.new)
+- [ ] **Lists**
+  - [ ] Use WASM GC array types
+  - [ ] Generate array operations (array.new, array.get, array.set)
+
+**Lower Priority** (optimizations and advanced features):
+- [ ] **Optimization passes**
+  - [ ] Dead code elimination
+  - [ ] Constant folding
+  - [ ] Common subexpression elimination
+  - [ ] Inlining small functions
+- [ ] **Export generation**
+  - [ ] Export functions with mangled names
+  - [ ] Generate getter functions for constant exports
+  - [ ] Handle module-level exports
+- [ ] **Better type handling**
+  - [ ] Use WASM GC reference types for complex types
+  - [ ] Implement runtime type tags for union types
+  - [ ] Generate type checking code for dynamic operations
+- [ ] **Debugging support**
+  - [ ] Generate DWARF debug info
+  - [ ] Source maps for WAT output
+  - [ ] Function name section for better stack traces
+- [ ] **Memory management**
+  - [ ] Integrate with WASM GC proposal
+  - [ ] Handle reference counting for shared values
+  - [ ] Optimize struct layout for cache efficiency
+
+**Testing Strategy**:
```

Comment:
I think at some point we should set up differential fuzz testing where we generate cadenza programs and evaluate them and compare the output to the wasm codegen. This will let us assert that our interpreter and codegen matches 100%. We'd need to add support for the interpreter to stop after a number of operations just so we don't get caught in an infinite loop. And honestly we probably need that anyway when we're executing things at compile time.